govcd/vm.go

```diff
@@ -164,53 +205,80 @@ func (vm *VM) ChangeCPUCountWithCore(virtualCpuCount int, coresPerSocket *int) (
-	}
+func (vm *VM) updateNicParameters(networks []map[string]interface{}, networkSection *types.NetworkConnectionSection) {
 
-	networkSection, err := vm.GetNetworkConnectionSection()
+	for tfNicSlot, network := range networks {
+		for loopIndex := range networkSection.NetworkConnection {
+			// Change network config only if we have the same virtual slot number as in .tf config
+			if tfNicSlot == networkSection.NetworkConnection[loopIndex].NetworkConnectionIndex {
 
-	// changes network config when only matches network name with provided one
-	for _, network := range networks {
-		for index, networkConnection := range networkSection.NetworkConnection {
-			if networkConnection.Network == network["orgnetwork"] { // network name are equal
 				// Determine what type of address is requested for the vApp
 				ipAllocationMode := types.IPAllocationModeNone
 				ipAddress := "Any"
 
-				// TODO: Review current behaviour of using DHCP when left blank
-				if ip == "dhcp" || network["ip"].(string) == "dhcp" {
+				var ipFieldString string
+
+				ipField, ipIsSet := network["ip"]
+				if ipIsSet {
+					ipFieldString = ipField.(string)
+				}
+
+				switch {
+				// TODO 3.0 remove from here when legacy `ip` and `network_name` attributes are removed
+				case ipIsSet && ipFieldString == "dhcp": // Legacy ip="dhcp" mode
 					ipAllocationMode = types.IPAllocationModeDHCP
-				} else if ip == "allocated" || network["ip"].(string) == "allocated" {
+				case ipIsSet && ipFieldString == "allocated": // Legacy ip="allocated" mode
 					ipAllocationMode = types.IPAllocationModePool
-				} else if ip == "none" || network["ip"].(string) == "none" {
+				case ipIsSet && ipFieldString == "none": // Legacy ip="none" mode
 					ipAllocationMode = types.IPAllocationModeNone
-				} else if ip != "" {
-					ipAllocationMode = types.IPAllocationModeManual
-					// TODO: Check a valid IP has been given
-					ipAddress = ip
-				} else if network["ip"].(string) != "" {
+
+				// Legacy ip="valid_ip" mode (currently it is hit by ip_allocation_mode=MANUAL as well)
+				case ipIsSet && net.ParseIP(ipFieldString) != nil:
 					ipAllocationMode = types.IPAllocationModeManual
-					// TODO: Check a valid IP has been given
-					ipAddress = network["ip"].(string)
-				} else if ip == "" {
+					ipAddress = ipFieldString
+				case ipIsSet && ipField != "": // Legacy ip="something_invalid" we default to DHCP. This is odd but backwards compatible.
 					ipAllocationMode = types.IPAllocationModeDHCP
+					// TODO 3.0 remove until here when legacy `ip` and `network_name` attributes are removed
+
+				case ipIsSet && net.ParseIP(ipFieldString) != nil && (network["ip_allocation_mode"].(string) == types.IPAllocationModeManual):
+					ipAllocationMode = types.IPAllocationModeManual
+					ipAddress = ipFieldString
+				default: // New functionality. IP was not set and we're defaulting to provided ip_allocation_mode (only manual requires the IP)
```

Comment:
What about "New **networks** functionality" to make sure it's clear we're referring to multi networks here.

Reply:
Changed